src/util.rs

```diff
@@ -0,0 +1,41 @@
+use std::ops::{Deref, DerefMut};
+
+use pyo3::prelude::*;
+use pyo3::types::{PyAny, PyList};
+
+/// A `Vec` that that can only be converted from a Python list.
+pub struct ListVec<T> {
+    inner: Vec<T>,
+}
+
+impl<T> ListVec<T> {
+    /// Get wrapped `Vec`.
+    pub fn into_inner(self) -> Vec<T> {
+        self.inner
+    }
+}
+
+impl<T> Deref for ListVec<T> {
+    type Target = Vec<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T> DerefMut for ListVec<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<'a, T> FromPyObject<'a> for ListVec<T>
+where
+    T: FromPyObject<'a>,
+{
+    fn extract(ob: &'a PyAny) -> PyResult<Self> {
```

Comment:
I am not aware of the performance implications of using python methods in rust land, but something like the method below would a) provide verbose errors (Currently, we get just `TypeError` without any message) and b) fail with everything that's not a list or tuple.

~~~rust
    fn extract(ob: &'a PyAny) -> PyResult<Self> {
        let gil = Python::acquire_gil();
        let py = gil.python();
        let is_list = py.is_instance::<PyList,_>(ob)?;
        let is_tuple = py.is_instance::<PyTuple,_>(ob)?;

        if ! (is_list || is_tuple) {
            Err(exceptions::TypeError::py_err("Invalid type! Can only construct a Vec from list or tuple!"))?
        } else {
            let inner : Vec<T> = if is_list {
                ob.downcast_ref::<PyList>()?.as_ref().extract()?
            } else {
                ob.downcast_ref::<PyTuple>()?.as_ref().extract()?
            };
            Ok(ListVec {inner})
        }
    }
~~~

Another no-gil method (I guess that's handled internally by pyo3 then?)  but kinda ugly method could:

~~~rust
    fn extract(ob: &'a PyAny) -> PyResult<Self> {
        let inner = match ob.downcast_ref::<PyList>() {
            Ok(list) => list.as_ref().extract()?,
            Err(_) => match ob.downcast_ref::<PyTuple>(){
                Ok(tuple) => tuple.as_ref().extract()?,
                Err(_) => Err(exceptions::TypeError::py_err("Invalid type! Can only extract a Vec from list or tuple!"))?
            }
        };
        Ok(ListVec {inner})
    }
~~~

I don't like the second since there may be other reasons for failing than `ob` not being a tuple / list.

Reply:
I don't want to accept tuples, since I do not really see the use case. It would just be weird to put sentences in a tuple.

I actually had a more detailed error first, but found some examples where other Python libraries just throw a `TypeError`. But I am fine with adding it back.

By the way, your second example can be written more clearly by e.g. using `Result::or`/`Result::or_else`.

Reply:
I added an error message.

Reply:
> I don't want to accept tuples, since I do not really see the use case. It would just be weird to put sentences in a tuple.

Sure, still a bit unexpected as a tuple provides what we need to convert it to a sentence. But I am also ok with forcing lists here.

Reply:
One thing I am not 100% happy with is that you cannot see which of the two arguments gives the `TypeError`. This could of course be done by taking `PyAny`s as arguments and checking per argument, but then the signature gets a bit uglier :/.